[Clang] [Sema] Don't crash on unexpanded pack in invalid block literal

Sirraide opened the pull request:
Consider #109148:
```c++
template <typename ...Ts>
void f() {
    [] {
        (^Ts);
    };
}
```

When we encounter `^Ts`, we try to parse a block and subsequently call `DiagnoseUnexpandedParameterPack()` (in `ActOnBlockArguments()`), which sees `Ts` and sets `ContainsUnexpandedParameterPack` to `true` in the `LambdaScopeInfo` of the enclosing lambda. However, the entire block is subsequently discarded entirely because it isn’t even syntactically well-formed. As a result, `ContainsUnexpandedParameterPack` is `true` despite the lambda’s body no longer containing any unexpanded packs, which causes an assertion the next time `DiagnoseUnexpandedParameterPack()` is called.

This pr moves handling of unexpanded parameter packs into `CapturingScopeInfo` instead so that the same logic is used for both blocks and lambdas. This fixes this issue since the `ContainsUnexpandedParameterPack` flag is now part of the block (and before that, its `CapturingScopeInfo`) and no longer affects the surrounding lambda directly when the block is parsed. Moreover, this change makes blocks actually usable with pack expansion.

This fixes #109148.

rjmccall commented:
I don't think there's a deep reason blocks and lambdas don't use quite the same scope mechanics in the compiler, so if you wanted to pursue that, it seems reasonable.  But this approach also seems viable.  I agree that removing the assertion would be the wrong thing to do, and I also don't see a reliable way to include the pack in the AST.

hokein commented:
The current solution seems reasonable to me.

> somehow including the malformed block in the AST (by wrapping it in a RecoveryExpr perhaps?)

`RecoveryExpr` is designed to preserve `Expr` nodes, so it can't be used to retain arbitrary AST nodes. In the case of `(^Ts)`, as far as I understand, `Ts` is parsed as a block-id, making RecoveryExpr inapplicable.

Sirraide commented:
> I don't think there's a deep reason blocks and lambdas don't use quite the same scope mechanics in the compiler, so if you wanted to pursue that, it seems reasonable. But this approach also seems viable.

Yeah, there is a comment about parameter packs in block arguments being a problem somewhere, but that’s from 2012, so that might no longer apply. I might look into this at some point if I have the time, but not for this pr (there are still other issues need to deal with first...)

cor3ntin commented:
> Side note: lambdas don’t suffer from this problem because they push a new LambdaScopeInfo, so if the lambda is discarded, the ‘enclosing lambda’ (which is just that very same lambda) is discarded along with it. I’m not sure making blocks work the same way would work (i.e. allowing unexpanded parameter packs if there is an enclosing lambda or block, but I’m assuming there is a reason why we don’t support that...), because currently, e.g. this doesn’t work at all:

I am not sure i understand that.

Imo the issue is that `DiagnoseUnexpandedParameterPacks` (where we set `LambdaScopeInfo::ContainsUnexpandedParameterPack` does not account for blocks - and probably should.

Ideally
 - DiagnoseUnexpandedParameterPacks would handle arbitrary capturing-scope thingies (statement expressions, lambdas, blocks, eventually do expressions, and other terrifying such constructs)
 - Once the innermost such construct is fully constructed, the `ContainsUnexpandedParameterPack` flag could bubble upward to the enclosing capturing scope

Sirraide commented:
> I am not sure i understand that.
>
> Imo the issue is that `DiagnoseUnexpandedParameterPacks` (where we set `LambdaScopeInfo::ContainsUnexpandedParameterPack` does not account for blocks - and probably should.
> 
> Ideally
> 
> * DiagnoseUnexpandedParameterPacks would handle arbitrary capturing-scope thingies (statement expressions, lambdas, blocks, eventually do expressions, and other terrifying such constructs)
> * Once the innermost such construct is fully constructed, the `ContainsUnexpandedParameterPack` flag could bubble upward to the enclosing capturing scope

Yeah, that’s also what I figured initially. I just assumed that there was *some* reason why blocks don’t or can’t work that way (I have never used Objective-C and I candidly don’t know a lot about blocks...), but it seems that there isn’t.

So if you think that this is a better idea, I can instead look into refactoring blocks to work the same way as lambdas wrt `DiagnoseUnexpandedParameterPacks()` (and I think that mainly just means moving some things from `LambdaScopeInfo` to `CapturingScopeInfo` or whatever the base class was called again and making `DiagnoseUnexpandedParameterPacks()` etc. use that instead).

Sirraide committed:
Undo previous changes, keeping the release note and tests

Sirraide committed:
Move unexpanded parameter handling into CapturingScopeInfo

Sirraide committed:
clang-format

Sirraide committed:
No longer throw away the block type if it contains unexpanded packs

zyn0217 commented on clang/lib/Sema/SemaTemplateVariadic.cpp:
```diff
@@ -262,20 +263,34 @@ namespace {
+      if (!Block->containsUnexpandedParameterPack())
+        return true;
+
+      bool WasInLambdaOrBlock = InLambdaOrBlock;
+      InLambdaOrBlock = true;
+
+      inherited::TraverseBlockExpr(Block);
+
+      InLambdaOrBlock = WasInLambdaOrBlock;
```
Maybe use llvm::SaveAndRestore?

Sirraide replied:
*That’s* what it was called. When I looked at the code that handles lambdas above I was like ‘don’t we have an RAII thingy for this?’ but I couldn’t remember what it was called, so thanks!

Sirraide committed:
[NFC] Use SaveAndRestore

rjmccall reviewed (approved):
Nice, this looks really clean.  Thanks for taking this on!

Sirraide committed:
Merge branch 'main' into block-unexpanded

Sirraide commented:
> C/C++ code formatter, clang-format found issues in your code

... I don’t think I changed anything about that include? Or did it just reorder them because I added one? :thinking:
